kvserver: use narrow checkpoints in consistency checker

pav-kv opened the pull request:
This commit modifies the consistency checker to save partial checkpoints
instead of full ones. The partial checkpoints contain the data for the
inconsistent range and its immediate left and right neighbour in the Store.
Only the inconsistent range's Replica is locked at the time of snapshotting the
store before creating the checkpoint, other ranges can be out of sync. They are
checkpointed too, to give more context when e.g. debugging non-trivial
split/merge scenarios.

Release note (ops change): In the rare event of Range inconsistency, the
consistency checker saves a storage checkpoint on each storage the Range
belongs to. Previously, this was a full checkpoint, so its cost could quickly
escalate on the nodes that went on running. This change makes the checkpoints
partial, i.e. they now only contain the relevant range and its neighbours. This
eliminates the time pressure on the cluster operator to remove the checkpoints.

Resolves #90543
Epic: none

pav-kv committed:
kvserver: use narrow checkpoints in consistency checker

This commit modifies the consistency checker to save partial checkpoints
instead of full ones. The partial checkpoints contain the data for the
inconsistent range and its immediate left and right neighbour in the Store.
Only the inconsistent range's Replica is locked at the time of snapshotting the
store before creating the checkpoint, other ranges can be out of sync. They are
checkpointed too, to give more context when e.g. debugging non-trivial
split/merge scenarios.

Release note (ops change): In the rare event of Range inconsistency, the
consistency checker saves a storage checkpoint on each storage the Range
belongs to. Previously, this was a full checkpoint, so its cost could quickly
escalate on the nodes that went on running. This change makes the checkpoints
partial, i.e. they now only contain the relevant range and its neighbours. This
eliminates the time pressure on the cluster operator to remove the checkpoints.

Epic: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/95963)
<!-- Reviewable:end -->

pavelkalinnikov commented on pkg/kv/kvserver/store.go:
```diff
@@ -2870,15 +2871,44 @@ func (s *Store) checkpointsDir() string {
+	if desc != nil {
+		// Find left and right neighbour of the range.
+		s.mu.RLock()
+		left := s.mu.replicasByKey.LookupPrecedingReplica(context.Background(), desc.StartKey).Desc()
+		var right *roachpb.RangeDescriptor
+		err := s.mu.replicasByKey.VisitKeyRange(context.Background(), desc.EndKey, roachpb.RKeyMax,
```
`LookupPrecedingReplica` finds a `Replica`, whereas the `VisitKeyRange` with the visitor written here find a `Replica` or a placeholder. We could unify this so that both `left` and `right` are `Replica`, or both are `Replica`/placeholder. Which way makes more sense?

Placeholders don't have much data, so could be skipped until we find a regular `Replica`. OTOH, if there are too many placeholders (or all of them), we would still checkpoint a large space.

pavelkalinnikov replied:
To guard against checkpointing too much, we can provide 3 `CheckpointSpan`s below, instead of a single wider one. The risk with one wider span is that the neighbouring ranges are not locked, so we may end up touching more ranges than intended.

erikgrinaker replied:
>LookupPrecedingReplica finds a Replica, whereas the VisitKeyRange with the visitor written here finds a Replica or a placeholder. We could unify this so that both left and right are Replica, or both are Replica/placeholder. Which way makes more sense?

We don't care about placeholders, I think. We can e.g. add a corresponding `LookupNextReplica` with the same behavior.

>To guard against checkpointing too much, we can provide 3 CheckpointSpans below, instead of a single wider one. The risk with one wider span is that the neighbouring ranges are not locked, so we may end up touching more ranges than intended.

Yeah, this can be pathological in certain extreme cases, e.g. where we recently removed a bunch of replicas but they haven't been compacted all the way down yet. But I suppose there's the same risk if we just merged a bunch of ranges into the inconsistent range.

I think we probably want to checkpoint the entire span between the neighbouring replicas, because any recently-removed ranges could be relevant to the inconsistency, and we'll just have to take the risk that they may contain more than 3 ranges worth of data. In most cases they wouldn't, and in any case it's a strict improvement over the current state -- we can reconsider if we do find any issues with it later.

pavelkalinnikov replied:
Done.

erikgrinaker reviewed (commented):
Would be worth having an integration test for this.

erikgrinaker commented on pkg/kv/kvserver/replica_consistency.go:
```diff
@@ -665,7 +673,14 @@ func (r *Replica) computeChecksumPostApply(
+		// few neighbouring replicas.
+		desc := &desc
+		if shouldFatal {
+			desc = nil
```
PS: I think we should take the same checkpoint across all nodes. On the terminated node we still have the entire database available if we need to look at anything else.

pavelkalinnikov replied:
We wouldn't be able to easily take the **same** checkpoint (i.e. with matching begin/end key), since different nodes may have different replica sets. But we can do the same algorithm: checkpoint the range and its neighbours.

erikgrinaker replied:
>We wouldn't be able to easily take the same checkpoint (i.e. with matching begin/end key), since different nodes may have different replica sets. But we can do the same algorithm: checkpoint the range and its neighbours.

Sure, that's fine -- this is mostly useful for Pebble-level stuff, where the LSM structure is going to be different across nodes anyway.

pavelkalinnikov replied:
Done. Using the same checkpoint algorithm regardless of the node termination.

erikgrinaker commented:
PPS: We should test that we can open the narrow checkpoint just like any other Pebble database as well, e.g. by running various `cockroach debug` commands on it.

RaduBerinde commented:
*[`pkg/kv/kvserver/store.go` line 2892 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95963#-NMhaxMVaqU5VOtEQDRm-r1-2892:-NMicdWA7idcepFVfwMj:bn9ndsc) ([raw file](https://github.com/cockroachdb/cockroach/blob/6daea4db70270ce04bf99ae6e51a07dc458d64bd/pkg/kv/kvserver/store.go#L2892)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

>LookupPrecedingReplica finds a Replica, whereas the VisitKeyRange with the visitor written here finds a Replica or a placeholder. We could unify this so that both left and right are Replica, or both are Replica/placeholder. Which way makes more sense?

We don't care about placeholders, I think. We can e.g. add a corresponding `LookupNextReplica` with the same behavior.

>To guard against checkpointing too much, we can provide 3 CheckpointSpans below, instead of a single wider one. The risk with one wider span is that the neighbouring ranges are not locked, so we may end up touching more ranges than intended.

Yeah, this can be pathological in certain extreme cases, e.g. where we recently removed a bunch of replicas but they haven't been compacted all the way down yet. But I suppose there's the same risk if we just merged a bunch of ranges into the inconsistent range.

I think we probably want to checkpoint the entire span between the neighbouring replicas, because any recently-removed ranges could be relevant to the inconsistency, and we'll just have to take the risk that they may contain more than 3 ranges worth of data. In most cases they wouldn't, and in any case it's a strict improvement over the current state -- we can reconsider if we do find any issues with it later.
</blockquote></details>

Another vote for a single span over the ranges. Note that the narrow checkpoints only exclude SST files that are fully outside of the spans; they do not "cut up" SST files that intersect the spans (and they don't trim WALs). So if we have ranges `[a c)`, `[f h)`, we may still see keys like `d`. That might become confusing when we use this for debugging.  We should take one continuous range that contains the areas of interest.

The exception of course would be if we want to also include something from the very beginning of the keyspace (the meta ranges?).


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95963)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jbowens, @pavelkalinnikov, @RaduBerinde, and @tbg)

___
*[`pkg/kv/kvserver/store.go` line 2892 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95963#-NMhaxMVaqU5VOtEQDRm-r1-2892:-NMif6eE107neoQPUTqR:buikwdn) ([raw file](https://github.com/cockroachdb/cockroach/blob/6daea4db70270ce04bf99ae6e51a07dc458d64bd/pkg/kv/kvserver/store.go#L2892)):*
>The exception of course would be if we want to also include something from the very beginning of the keyspace (the meta ranges?).

The meta ranges likely won't be on this node in larger clusters, and it isn't clear that they'd be helpful anyway. But I do think we want to get the range-local keyspans, both replicated and unreplicated, and those will have to be separate spans since they're prefixed by range ID.

This might be confusing, as your say, since we'll see various parts of the surrounding ranges at various LSM levels, but I think in almost all cases we'll only be interested in the user keyspan anyway.


<!-- Sent from Reviewable.io -->

pav-kv reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95963)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker, @jbowens, @RaduBerinde, and @tbg)

___
*[`pkg/kv/kvserver/store.go` line 2892 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95963#-NMhaxMVaqU5VOtEQDRm-r1-2892:-NMj8GhI6KXbHrx9fVzc:b-gtxwlt) ([raw file](https://github.com/cockroachdb/cockroach/blob/6daea4db70270ce04bf99ae6e51a07dc458d64bd/pkg/kv/kvserver/store.go#L2892)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

>The exception of course would be if we want to also include something from the very beginning of the keyspace (the meta ranges?).

The meta ranges likely won't be on this node in larger clusters, and it isn't clear that they'd be helpful anyway. But I do think we want to get the range-local keyspans, both replicated and unreplicated, and those will have to be separate spans since they're prefixed by range ID.

This might be confusing, as your say, since we'll see various parts of the surrounding ranges at various LSM levels, but I think in almost all cases we'll only be interested in the user keyspan anyway.
</blockquote></details>

I'm also inclined to make it a single span. One complication here is that one range already consists of multiple spans. See `rditer.makeAllKeySpans`, I counted 6 spans. Now, if we have 3 ranges 6 spans each, we can either use the 18 spans as they are, or merge them into 6 wide spans. I'm in favour of the latter approach, but it looks like we would need cross-rangeID span helpers for that.


<!-- Sent from Reviewable.io -->

pav-kv reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95963)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jbowens and @tbg)

___
*[`pkg/kv/kvserver/store.go` line 2892 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95963#-NMhaxMVaqU5VOtEQDRm-r1-2892:-NMjBF3jEwrqU3IaOSuD:b-86okv5) ([raw file](https://github.com/cockroachdb/cockroach/blob/6daea4db70270ce04bf99ae6e51a07dc458d64bd/pkg/kv/kvserver/store.go#L2892)):*
<details><summary><i>Previously, pavelkalinnikov (Pavel Kalinnikov) wrote…</i></summary><blockquote>

I'm also inclined to make it a single span. One complication here is that one range already consists of multiple spans. See `rditer.makeAllKeySpans`, I counted 6 spans. Now, if we have 3 ranges 6 spans each, we can either use the 18 spans as they are, or merge them into 6 wide spans. I'm in favour of the latter approach, but it looks like we would need cross-rangeID span helpers for that.
</blockquote></details>

@tbg Do you think we could adapt `rditer/select.go` to work across range IDs?

Right now we have:
```golang
func Select(rangeID roachpb.RangeID, opts SelectOpts) []roachpb.Span
```

But it seems easy to extend it to:
```golang
func SelectMulti(begin, end roachpb.RangeID, opts SelectOpts) []roachpb.Span
```

It would use the starting points of the `begin` range, and end points of the `end` range. A little complication is that the caller must correctly compute the user keys span in `opts`. For a single range it's easy: take the range descriptor's span; for a multi-range case the caller should take the first key of the left range, and the end key of the last range.


<!-- Sent from Reviewable.io -->

pav-kv reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95963)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jbowens and @tbg)

___
*[`pkg/kv/kvserver/store.go` line 2892 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95963#-NMhaxMVaqU5VOtEQDRm-r1-2892:-NMjDE_k4LwsMhboF4l2:bqrkccz) ([raw file](https://github.com/cockroachdb/cockroach/blob/6daea4db70270ce04bf99ae6e51a07dc458d64bd/pkg/kv/kvserver/store.go#L2892)):*
<details><summary><i>Previously, pavelkalinnikov (Pavel Kalinnikov) wrote…</i></summary><blockquote>

@tbg Do you think we could adapt `rditer/select.go` to work across range IDs?

Right now we have:
```golang
func Select(rangeID roachpb.RangeID, opts SelectOpts) []roachpb.Span
```

But it seems easy to extend it to:
```golang
func SelectMulti(begin, end roachpb.RangeID, opts SelectOpts) []roachpb.Span
```

It would use the starting points of the `begin` range, and end points of the `end` range. A little complication is that the caller must correctly compute the user keys span in `opts`. For a single range it's easy: take the range descriptor's span; for a multi-range case the caller should take the first key of the left range, and the end key of the last range.
</blockquote></details>

My mistake, I think we can't assume that consecutive ranges according to the key are consecutive w.r.t. range IDs. So, for all rangeID-based keys, we need to take all the spans as is. It's only the user key spans that we can (and should) merge.


<!-- Sent from Reviewable.io -->

pav-kv reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95963)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jbowens and @tbg)

___
*[`pkg/kv/kvserver/store.go` line 2892 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95963#-NMhaxMVaqU5VOtEQDRm-r1-2892:-NMjEqJIEgqFU7Nyudc8:b-8z3gx0) ([raw file](https://github.com/cockroachdb/cockroach/blob/6daea4db70270ce04bf99ae6e51a07dc458d64bd/pkg/kv/kvserver/store.go#L2892)):*
<details><summary><i>Previously, pavelkalinnikov (Pavel Kalinnikov) wrote…</i></summary><blockquote>

My mistake, I think we can't assume that consecutive ranges according to the key are consecutive w.r.t. range IDs. So, for all rangeID-based keys, we need to take all the spans as is. It's only the user key spans that we can (and should) merge.
</blockquote></details>

Hm, but then we're limited to only finding bugs that affect the user keyspace. What if the bug is in the rangeID-based space? Wouldn't we want the LSMs touching neighbouring range IDs included as well?

@erikgrinaker Do you remember which part of the keyspace contained that Pebble bug? Could it manifest itself in other parts?


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95963)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jbowens, @pavelkalinnikov, and @tbg)

___
*[`pkg/kv/kvserver/store.go` line 2892 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95963#-NMhaxMVaqU5VOtEQDRm-r1-2892:-NMjU6tQ-5aB7xncIm4n:b-esigi1) ([raw file](https://github.com/cockroachdb/cockroach/blob/6daea4db70270ce04bf99ae6e51a07dc458d64bd/pkg/kv/kvserver/store.go#L2892)):*
> My mistake, I think we can't assume that consecutive ranges according to the key are consecutive w.r.t. range IDs. So, for all rangeID-based keys, we need to take all the spans as is. It's only the user key spans that we can (and should) merge.

Correct, they'll often not have consecutive range IDs.

>What if the bug is in the rangeID-based space? Wouldn't we want the LSMs touching neighbouring range IDs included as well?

Yeah, may as well grab it just in case. For the range-local keyspace let's extend the spans to the neighbouring ranges by range ID, while for the user keyspace we'll extend it to the neighbouring ranges by key.

One could argue that we should also grab the range-local spans for the neighbouring ranges by user key, but I don't think we need to overcomplicate this. I think what we mostly care about is the inconsistent range itself and the surrounding LSM contents, and so let's just grab that.

>Do you remember which part of the keyspace contained that Pebble bug? Could it manifest itself in other parts?

The user keyspace.

It could manifest in the range-local keyspace too, since we do use Pebble range tombstones there, but I think they're much less likely to span ranges given the structure of the keyspace there (it typically has many different prefixes for different types of data).


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
There's a `roachtest/inconsistency` roachtest that we might want to spruce up a bit.

> Reviewed 5 of 5 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95963#-:-NMnSfRf0ipLK3IUSSM8:b-gvad0q)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker, @jbowens, @pavelkalinnikov, and @RaduBerinde)

___
*[`pkg/kv/kvserver/replica_consistency.go` line 778 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/95963#-NMnQhqG8yNulanV0O-k:-NMnQhqG8yNulanV0O-l:b-i7bmbw) ([raw file](https://github.com/cockroachdb/cockroach/blob/16374ce7c718e39b87bf29df80ef73dc0adfc014/pkg/kv/kvserver/replica_consistency.go#L778)):*
> ```Go
> compacted, which leads to checkpoints becoming a full copy of a past state. This
> is not time critical because the checkpoints are localized to a single / few
> range(s), and over time may grow to consume only O(GB) disk space. There is no
> ```

The range size is configurable, and also we're starting to talk about larger range sizes. You could make this a bit more general and say that it's `O(RangeSize)`.
___
*[`pkg/kv/kvserver/store.go` line 2892 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95963#-NMhaxMVaqU5VOtEQDRm-r1-2892:-NMnR-u571X-RvLdnW7N:bdua50y) ([raw file](https://github.com/cockroachdb/cockroach/blob/6daea4db70270ce04bf99ae6e51a07dc458d64bd/pkg/kv/kvserver/store.go#L2892)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

> My mistake, I think we can't assume that consecutive ranges according to the key are consecutive w.r.t. range IDs. So, for all rangeID-based keys, we need to take all the spans as is. It's only the user key spans that we can (and should) merge.

Correct, they'll often not have consecutive range IDs.

>What if the bug is in the rangeID-based space? Wouldn't we want the LSMs touching neighbouring range IDs included as well?

Yeah, may as well grab it just in case. For the range-local keyspace let's extend the spans to the neighbouring ranges by range ID, while for the user keyspace we'll extend it to the neighbouring ranges by key.

One could argue that we should also grab the range-local spans for the neighbouring ranges by user key, but I don't think we need to overcomplicate this. I think what we mostly care about is the inconsistent range itself and the surrounding LSM contents, and so let's just grab that.

>Do you remember which part of the keyspace contained that Pebble bug? Could it manifest itself in other parts?

The user keyspace.

It could manifest in the range-local keyspace too, since we do use Pebble range tombstones there, but I think they're much less likely to span ranges given the structure of the keyspace there (it typically has many different prefixes for different types of data).
</blockquote></details>

Determining a single `RSpan` over which to checkpoint the key-prefixed data makes sense to me. Space that has a placeholder on it ought to be empty anyway. So I think we should:

- find left and right neighboring range, note the left start key, the right end key, and both range ids
- rspan=`[left start key, right end key)`
- `spans = append(rditer.Select(rangeID, ReplicatedBySpan:rspan, ReplicatedByRangeID:true), rditer.Select(leftRangeID, ReplicatedByRangeID:true), rditer.Select(rightRangeID, ReplicatedByRangeID:true)`

In words, we pretend that our range has the wider key span and checkpoint that (this is a couple of spans under the hood due to local-key and local-rangeid addressed keys), but in addition include the two replicated rangeid-based key spans for the left and right neighbor.

For debugging purposes we should also throw in the rangeid-unreplicated key spans, i.e. set the corresponding bool to true in `spans = ` above for all `Select` invocations.
___
*[`pkg/kv/kvserver/store.go` line 2911 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/95963#-NMnSIh_3vlpy61nt6wS:-NMnSIh_3vlpy61nt6wT:b-sh57bq) ([raw file](https://github.com/cockroachdb/cockroach/blob/16374ce7c718e39b87bf29df80ef73dc0adfc014/pkg/kv/kvserver/store.go#L2911)):*
> ```Go
> 		spans = append(spans, span)
> 	}
> 	if err := s.engine.CreateCheckpoint(checkpointDir, spans); err != nil {
> ```

It would be good to mark the checkpoint as partial and perhaps encode (in a text file?) what the start and end key and rangeIDs were that are guaranteed to be covered.


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95963)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jbowens, @pavelkalinnikov, @RaduBerinde, and @tbg)

___
*[`pkg/kv/kvserver/store.go` line 2892 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95963#-NMhaxMVaqU5VOtEQDRm-r1-2892:-NMnhXQB0e-WUp_OXO35:b-fdxvt4) ([raw file](https://github.com/cockroachdb/cockroach/blob/6daea4db70270ce04bf99ae6e51a07dc458d64bd/pkg/kv/kvserver/store.go#L2892)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Determining a single `RSpan` over which to checkpoint the key-prefixed data makes sense to me. Space that has a placeholder on it ought to be empty anyway. So I think we should:

- find left and right neighboring range, note the left start key, the right end key, and both range ids
- rspan=`[left start key, right end key)`
- `spans = append(rditer.Select(rangeID, ReplicatedBySpan:rspan, ReplicatedByRangeID:true), rditer.Select(leftRangeID, ReplicatedByRangeID:true), rditer.Select(rightRangeID, ReplicatedByRangeID:true)`

In words, we pretend that our range has the wider key span and checkpoint that (this is a couple of spans under the hood due to local-key and local-rangeid addressed keys), but in addition include the two replicated rangeid-based key spans for the left and right neighbor.

For debugging purposes we should also throw in the rangeid-unreplicated key spans, i.e. set the corresponding bool to true in `spans = ` above for all `Select` invocations.

</blockquote></details>

Yep. Note that the rangeID neighbours will often be different ranges than the user key neighbours. We don't want to conflate them, since we could otherwise checkpoint the entire range-local span, which can be pretty large. In other words, if we have an inconsistent range with the following surrounding ranges on that node:

* [a-b): range 8
* [d-f): range 5 (inconsistent)
* [k-m): range 2

The node has replicas for ranges 2, 3, 5, 6, 8.

Then we want to checkpoint the user keyspan [a-m), and the range-local keyspans for ranges [3-6].


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95963)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jbowens, @pavelkalinnikov, @RaduBerinde, and @tbg)

___
*[`pkg/kv/kvserver/store.go` line 2892 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95963#-NMhaxMVaqU5VOtEQDRm-r1-2892:-NMnjlezEYubEWnBkMAj:b-ooxo8p) ([raw file](https://github.com/cockroachdb/cockroach/blob/6daea4db70270ce04bf99ae6e51a07dc458d64bd/pkg/kv/kvserver/store.go#L2892)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Yep. Note that the rangeID neighbours will often be different ranges than the user key neighbours. We don't want to conflate them, since we could otherwise checkpoint the entire range-local span, which can be pretty large. In other words, if we have an inconsistent range with the following surrounding ranges on that node:

* [a-b): range 8
* [d-f): range 5 (inconsistent)
* [k-m): range 2

The node has replicas for ranges 2, 3, 5, 6, 8.

Then we want to checkpoint the user keyspan [a-m), and the range-local keyspans for ranges [3-6].
</blockquote></details>

Taking into account whether the local keyspans are prefixed by key or range ID, of course.


<!-- Sent from Reviewable.io -->

pav-kv reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95963)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker, @jbowens, @RaduBerinde, and @tbg)

___
*[`pkg/kv/kvserver/store.go` line 2892 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95963#-NMhaxMVaqU5VOtEQDRm-r1-2892:-NMntqQs37ylzjXTE3nd:b49iv5m) ([raw file](https://github.com/cockroachdb/cockroach/blob/6daea4db70270ce04bf99ae6e51a07dc458d64bd/pkg/kv/kvserver/store.go#L2892)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Taking into account whether the local keyspans are prefixed by key or range ID, of course.
</blockquote></details>

SGTM, thanks for the suggestions. WIP.
___
*[`pkg/kv/kvserver/store.go` line 2911 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/95963#-NMnSIh_3vlpy61nt6wS:-NMnu0K69yzoZ-QzKkDT:b-dcisjq) ([raw file](https://github.com/cockroachdb/cockroach/blob/16374ce7c718e39b87bf29df80ef73dc0adfc014/pkg/kv/kvserver/store.go#L2911)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

It would be good to mark the checkpoint as partial and perhaps encode (in a text file?) what the start and end key and rangeIDs were that are guaranteed to be covered.
</blockquote></details>

I agree. I was thinking at least including this information in the log/prevent-startup-file message. Do you think it should a separate file though?


<!-- Sent from Reviewable.io -->

jbowens reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95963)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker, @pavelkalinnikov, @RaduBerinde, and @tbg)

___
*[`pkg/kv/kvserver/store.go` line 2911 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/95963#-NMnSIh_3vlpy61nt6wS:-NMofWw3BobVBmO30yty:bqhyhi2) ([raw file](https://github.com/cockroachdb/cockroach/blob/16374ce7c718e39b87bf29df80ef73dc0adfc014/pkg/kv/kvserver/store.go#L2911)):*
<details><summary><i>Previously, pavelkalinnikov (Pavel Kalinnikov) wrote…</i></summary><blockquote>

I agree. I was thinking at least including this information in the log/prevent-startup-file message. Do you think it should a separate file though?
</blockquote></details>

+1 to encoding the spans somewhere: the Pebble checkpoint won't encode any information about which spans were checkpointed, and it won't be possible to tell which regions of the keyspace should be consistent. I could imagine looking at one of these partial checkpoints and getting confused because there exist keys outside the checkpointed spans that are inconsistent.


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[`pkg/kv/kvserver/store.go` line 2911 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/95963#-NMnSIh_3vlpy61nt6wS:-NMogtLZ4lphRqIRe_ov:b-rhyqwv) ([raw file](https://github.com/cockroachdb/cockroach/blob/16374ce7c718e39b87bf29df80ef73dc0adfc014/pkg/kv/kvserver/store.go#L2911)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

+1 to encoding the spans somewhere: the Pebble checkpoint won't encode any information about which spans were checkpointed, and it won't be possible to tell which regions of the keyspace should be consistent. I could imagine looking at one of these partial checkpoints and getting confused because there exist keys outside the checkpointed spans that are inconsistent.
</blockquote></details>

I think it would be a good idea to include this information in the checkpoint's manifest..  And have any tools you use warn you about it if these are set.. I'd need to update the manifest format though.


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[`pkg/kv/kvserver/store.go` line 2911 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/95963#-NMnSIh_3vlpy61nt6wS:-NMojop6DgJEnEFSQFPy:b-sa4gpy) ([raw file](https://github.com/cockroachdb/cockroach/blob/16374ce7c718e39b87bf29df80ef73dc0adfc014/pkg/kv/kvserver/store.go#L2911)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I think it would be a good idea to include this information in the checkpoint's manifest..  And have any tools you use warn you about it if these are set.. I'd need to update the manifest format though.
</blockquote></details>

Filed https://github.com/cockroachdb/pebble/issues/2285


<!-- Sent from Reviewable.io -->

pavelkalinnikov commented on pkg/kv/kvserver/replica_consistency.go:
```diff
@@ -759,11 +767,16 @@ Checkpoints are created on each node/store hosting this range, to help
+On the surviving nodes, checkpoints are partial, i.e. contain only the data
+specific to the inconsistent range, and possibly its neighbouring ranges.
+
+The storage checkpoint directories MUST be deleted or moved away eventually, on
```
Done.

pavelkalinnikov commented on pkg/kv/kvserver/store.go:
```diff
@@ -2870,15 +2871,44 @@ func (s *Store) checkpointsDir() string {
+// Creates the checkpoint of the entire storage if the provided range descriptor
+// is nil. Otherwise, creates a "narrow" checkpoint that includes the provided
+// range, and possibly a few neighbouring ranges.
+func (s *Store) checkpoint(tag string, desc *roachpb.RangeDescriptor) (string, error) {
```
Done.

pavelkalinnikov commented on pkg/kv/kvserver/store.go:
```diff
@@ -2870,15 +2871,44 @@ func (s *Store) checkpointsDir() string {
+			return "", err
+		}
+
+		span := pebble.CheckpointSpan{Start: desc.StartKey, End: desc.EndKey}
```
Done.

pav-kv committed:
roachtest,kvserver: brush up consistency check tests

Release note: none
Epic: none

tbg reviewed (commented):
> Dismissed @erikgrinaker from 2 discussions.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95963)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker, @jbowens, @pavelkalinnikov, and @RaduBerinde)

___
*[`pkg/kv/kvserver/replica_consistency.go` line 639 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95963#-NN2Z5VN2S53aybXx7vX:-NN2Z5VN2S53aybXx7vY:bhevnlh) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b84f2aef227fbdbf789074f841c34fe5f176392/pkg/kv/kvserver/replica_consistency.go#L639)):*
> ```Go
> }
> 
> func (r *Replica) checkpointSpans() []roachpb.Span {
> ```

`func (s *Store) checkpointSpans(desc *roachpb.RangeDescriptor) []roachpb.Span`
___
*[`pkg/kv/kvserver/store.go` line 2892 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95963#-NMhaxMVaqU5VOtEQDRm-r1-2892:-NN2ZSbe6gZQiOrXKsUY:bhesjgw) ([raw file](https://github.com/cockroachdb/cockroach/blob/6daea4db70270ce04bf99ae6e51a07dc458d64bd/pkg/kv/kvserver/store.go#L2892)):*
<details><summary><i>Previously, pavelkalinnikov (Pavel Kalinnikov) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Just to double check, it looks like what was done finds the left and right "desc" neighbor, and then takes their rangeIDs only. I think Erik at least at one point suggested that we should also look for the "rangeID" neighbors and add those. (This would require basically creating the unreplicated rangeID prefix for the current desc, then searching backwards (thus discovering the "largest smallest" other rangeID) and doing something similar in the other direction as well). I'm fine with the current code, but just checking in that Erik is ok with it as well.
___
*[`pkg/kv/kvserver/store.go` line 2911 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/95963#-NMnSIh_3vlpy61nt6wS:-NN2YAvzBwKGpShZ3RZd:b51hhyc) ([raw file](https://github.com/cockroachdb/cockroach/blob/16374ce7c718e39b87bf29df80ef73dc0adfc014/pkg/kv/kvserver/store.go#L2911)):*
> I agree. I was thinking at least including this information in the log/prevent-startup-file message. Do you think it should a separate file though?

That's good for now, if the pebble folks want to improve on that by embedding it in the checkpoints that would be even better, as we don't always get the prevented startup file sent to us immediately.

I would prefer if in this PR we also "just" put a text file listing them into the checkpoint directory, basically a shortcut to the manifest, which has pretty printing etc.

The manifest doesn't know about CRDB ranges, so there'll be a loss of fidelity. I'd want that text file regardless. The text file could just contain the part of the prevented startup file that lists that information. (We'd want it on all checkpoints though).


<!-- Sent from Reviewable.io -->

pav-kv reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95963)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker, @herkolategan, @jbowens, @RaduBerinde, @srosenberg, and @tbg)

___
*[`pkg/kv/kvserver/store.go` line 2892 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95963#-NMhaxMVaqU5VOtEQDRm-r1-2892:-NN2e-aj98tpv1PW7rY6:bql3ebc) ([raw file](https://github.com/cockroachdb/cockroach/blob/6daea4db70270ce04bf99ae6e51a07dc458d64bd/pkg/kv/kvserver/store.go#L2892)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Just to double check, it looks like what was done finds the left and right "desc" neighbor, and then takes their rangeIDs only. I think Erik at least at one point suggested that we should also look for the "rangeID" neighbors and add those. (This would require basically creating the unreplicated rangeID prefix for the current desc, then searching backwards (thus discovering the "largest smallest" other rangeID) and doing something similar in the other direction as well). I'm fine with the current code, but just checking in that Erik is ok with it as well.
</blockquote></details>

Right, this isn't the final version, I just did the easy by-key part for now, and testing things. I will add by-rangeID neighbours shortly.


<!-- Sent from Reviewable.io -->

pav-kv committed:
roachtest: inspect checkpoint in consistency check test

This commit modifies the consistency check failure roachtest to verify that all
the nodes of an inconsistent range create a checkpoint that can be opened with
tooling.

Release note: none
Epic: none

pav-kv reviewed (commented):
Done. Modified the integration test, and also tested manually on a larger cluster (so that the checkpoint is indeed partial).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95963#-:-NN7JY81BJNCsyxilC3s:bhssgwn)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker, @herkolategan, @srosenberg, and @tbg)

___
*[`pkg/kv/kvserver/replica_consistency.go` line 778 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/95963#-NMnQhqG8yNulanV0O-k:-NN2vB2n09oatQntiDvA:bo63l53) ([raw file](https://github.com/cockroachdb/cockroach/blob/16374ce7c718e39b87bf29df80ef73dc0adfc014/pkg/kv/kvserver/replica_consistency.go#L778)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

The range size is configurable, and also we're starting to talk about larger range sizes. You could make this a bit more general and say that it's `O(RangeSize)`.
</blockquote></details>

Not mentioning the size anymore.
___
*[`pkg/kv/kvserver/replica_consistency.go` line 639 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95963#-NN2Z5VN2S53aybXx7vX:-NN2wLxkE-3fWBkkdN-D:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b84f2aef227fbdbf789074f841c34fe5f176392/pkg/kv/kvserver/replica_consistency.go#L639)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

`func (s *Store) checkpointSpans(desc *roachpb.RangeDescriptor) []roachpb.Span`
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/store.go` line 2911 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/95963#-NMnSIh_3vlpy61nt6wS:-NN7JqsmB9hpockz8xNH:b-fhri8f) ([raw file](https://github.com/cockroachdb/cockroach/blob/16374ce7c718e39b87bf29df80ef73dc0adfc014/pkg/kv/kvserver/store.go#L2911)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

> I agree. I was thinking at least including this information in the log/prevent-startup-file message. Do you think it should a separate file though?

That's good for now, if the pebble folks want to improve on that by embedding it in the checkpoints that would be even better, as we don't always get the prevented startup file sent to us immediately.

I would prefer if in this PR we also "just" put a text file listing them into the checkpoint directory, basically a shortcut to the manifest, which has pretty printing etc.

The manifest doesn't know about CRDB ranges, so there'll be a loss of fidelity. I'd want that text file regardless. The text file could just contain the part of the prevented startup file that lists that information. (We'd want it on all checkpoints though).
</blockquote></details>

Added a `checkpoint.txt` file in the Pebble's checkpoint directory, and a TODO pointing at the issue which will extend the manifest.


<!-- Sent from Reviewable.io -->

pav-kv reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95963)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker, @herkolategan, @srosenberg, and @tbg)

___
*[`pkg/kv/kvserver/store.go` line 2892 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95963#-NMhaxMVaqU5VOtEQDRm-r1-2892:-NN7ws4n1HN2pEnQRhZB:bcskzc8) ([raw file](https://github.com/cockroachdb/cockroach/blob/6daea4db70270ce04bf99ae6e51a07dc458d64bd/pkg/kv/kvserver/store.go#L2892)):*
<details><summary><i>Previously, pavelkalinnikov (Pavel Kalinnikov) wrote…</i></summary><blockquote>

Right, this isn't the final version, I just did the easy by-key part for now, and testing things. I will add by-rangeID neighbours shortly.
</blockquote></details>

Added the rangeID neighbours too.


<!-- Sent from Reviewable.io -->

pav-kv committed:
kvserver: add LookupNextReplica test

Release note: none
Epic: none

pavelkalinnikov commented:
@erikgrinaker @tbg This is ready for another round.

pavelkalinnikov commented on pkg/kv/kvserver/store.go:
```diff
@@ -2870,18 +2871,82 @@ func (s *Store) checkpointsDir() string {
+	if left != nil {
+		userKeys.Key = left.Desc().StartKey
+		if id := left.RangeID; id < prevID || id > nextID {
+			spans = append(spans, rditer.Select(left.RangeID, byRangeID)...)
```
s/left.RangeID/id/

pavelkalinnikov commented on pkg/kv/kvserver/store.go:
```diff
@@ -2870,18 +2871,82 @@ func (s *Store) checkpointsDir() string {
+	if right != nil {
+		userKeys.EndKey = right.Desc().EndKey
+		if id := right.RangeID; id < prevID || id > nextID {
+			spans = append(spans, rditer.Select(right.RangeID, byRangeID)...)
```
s/right.RangeID/id/

tbg reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with exception of the case in which we have neighboring rangeIDs but no neighboring inited ranges. It sounds like we're not collecting the neighboring rangeID spans in the checkpoint then. Either explain it in a comment or fix it, depending on it's broken. After that good to merge.

Do we have coverage that the checkpoint contains ... the right thing? If we passed the wrong spans to pebble or pebble would just create an empty checkpoint, would we find out anywhere? If not, let's add... some testing at least. Unit testing of `checkpointSpans` would seem appropriate for example, or we do some additional verification in `TestCheckConsistencyInconsistent`?

> Reviewed 1 of 5 files at r3, 1 of 3 files at r4, 5 of 5 files at r6, 2 of 2 files at r7, 1 of 1 files at r8, 1 of 1 files at r9, all commit messages.
> Dismissed @RaduBerinde and @jbowens from a discussion.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95963#-:-NNH542t9dDX--HV-T-n:bmh0lv8)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @erikgrinaker, @herkolategan, @pavelkalinnikov, and @srosenberg)

___
*[`pkg/cmd/roachtest/tests/inconsistency.go` line 145 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/95963#-NNH3wjP9r3y10QzATV5:-NNH3wjP9r3y10QzATV6:b6f4pxy) ([raw file](https://github.com/cockroachdb/cockroach/blob/23be73b40d85170c21d92123a4a527f6fdac1a4e/pkg/cmd/roachtest/tests/inconsistency.go#L145)):*
> ```Go
> 		// multiple if multiple consistency checks fail in close succession), and
> 		// puts spans information into the checkpoint.txt file in it.
> 		c.Run(ctx, c.Node(n), "find {store-dir}/auxiliary/checkpoints/r1_at_*/checkpoint.txt")
> ```

Heh, I thought this was a bogus invocation because the `*` looks like it would get shell expanded, but I think I added proper escaping here at some point. Still, this would look less weird as

```go
find {store-dir}/auxiliary/checkpoints -name 'checkpoint.txt'
```

I cede the loss of coverage of the `r1_` match.

Also, this would be the other approach:

```
		det, err := c.RunWithDetailsSingleNode()
		det.Stdout <-- has the path and can be validated/used for the subsequend cmds
```

I'm fine with the current code as-is just as long as it works.
___
*[`pkg/kv/kvserver/store.go` line 2923 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/95963#-NNH2g5SBYj6lCI0sGJf:-NNH2g5SBYj6lCI0sGJg:bfmt49) ([raw file](https://github.com/cockroachdb/cockroach/blob/85ae185ced438b5e4761aa87e9f0d4a7fb1e7049/pkg/kv/kvserver/store.go#L2923)):*
> ```Go
> 	if left != nil {
> 		userKeys.Key = left.Desc().StartKey
> 		if id := left.RangeID; id < prevID || id > nextID {
> ```

So if we don't hit this condition, we don't collect the neighboring rangeID spans? Ditto for the right branch.
___
*[`pkg/kv/kvserver/store_replica_btree_test.go` line 126 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/95963#-NNH51825q56Ff4BCZUr:-NNH51825q56Ff4BCZUs:b-r60t5m) ([raw file](https://github.com/cockroachdb/cockroach/blob/198090b2fed2c62d14cf63a8a333742eba617884/pkg/kv/kvserver/store_replica_btree_test.go#L126)):*
> ```Go
> 		{"a", nil, repl2},
> 		{"aa", nil, repl3},
> 		{"b", repl2, repl3},
> ```

Nice, thanks for adding the test.


<!-- Sent from Reviewable.io -->

pav-kv reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95963)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @erikgrinaker, @herkolategan, and @srosenberg)

___
*[`pkg/kv/kvserver/store.go` line 2923 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/95963#-NNH2g5SBYj6lCI0sGJf:-NNaBnH586HFh2_66-kh:bcp5k39) ([raw file](https://github.com/cockroachdb/cockroach/blob/85ae185ced438b5e4761aa87e9f0d4a7fb1e7049/pkg/kv/kvserver/store.go#L2923)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

So if we don't hit this condition, we don't collect the neighboring rangeID spans? Ditto for the right branch.
</blockquote></details>

We do collect them above where we take by-rangeID spans covering all of [prevID, nextID]. This check here skips adding spans for an ID that is in the already covered [prevID, nextID].


<!-- Sent from Reviewable.io -->

pav-kv reviewed (commented):
Yep, a unit-test for `checkpointSpans` would be nicest, I think. Will do.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95963#-:-NNaCa3fCBZjFNuEmwQm:baphdal)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @erikgrinaker, @herkolategan, and @srosenberg)



<!-- Sent from Reviewable.io -->

pav-kv committed:
kvserver: test Store.checkpointSpans

Release note: none
Epic: none

pav-kv reviewed (commented):
Add `TestStoreCheckpointSpans`. PTAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95963#-:-NNake6L11cREWK7FIog:b-tm2x9h)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @erikgrinaker, @herkolategan, @srosenberg, and @tbg)



<!-- Sent from Reviewable.io -->

pavelkalinnikov commented:
@erikgrinaker wanna take a look at which spans we checkpoint before I merge? See `Store.checkpointSpans` function and test.

erikgrinaker commented:
Yeah, I'll have a look.

tbg reviewed (approved):
> Reviewed 6 of 6 files at r10, 2 of 2 files at r11, 1 of 1 files at r12, 1 of 1 files at r13, 1 of 1 files at r14, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95963)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @erikgrinaker, @herkolategan, and @srosenberg)



<!-- Sent from Reviewable.io -->

erikgrinaker commented on pkg/kv/kvserver/store.go:
```diff
@@ -2879,18 +2880,82 @@ func (s *Store) checkpointsDir() string {
+	// Cover all range IDs (prevID, desc.RangeID, nextID) using continuous spans,
+	// one for replicated and one for unreplicated keyspace.
+	// TODO(tbg): make an rditer.Select-like helper for this.
+	spans := []roachpb.Span{{
```
Let's use `MakeRangeIDPrefix` -- this covers all infixes (both replicated, unreplicated, and any others that we might add), and is also simpler.

pavelkalinnikov replied:
Nice idea, thanks! Done.

pavelkalinnikov replied:
@tbg FYI, this ended up more convenient than leverating `rditer.Select` specifying both replicated and unreplicated by-rangeID space.

erikgrinaker reviewed (approved):
LGTM, thanks!

pavelkalinnikov commented:
bors r=tbg,erikgrinaker
